### PR TITLE
fixed SDXL text encoder training bug #5016

### PR DESCRIPTION
**Pull request by shyammarjit**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 1]
```diff
@@ -1066,6 +1066,11 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
         if args.train_text_encoder:
             text_encoder_one.train()
             text_encoder_two.train()
+            
+            # set top parameter requires_grad = True for gradient checkpointing works
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:1070]: I don't fully understand why we add these changes. Can you explain a bit @shyammarjit ? How does this solve an error / warning with gradient checkpointing?

Comment (shyammarjit) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:1070]: @patrickvonplaten please have a look at this issue https://github.com/huggingface/diffusers/issues/5016 and its follow-up issue at https://github.com/kohya-ss/sd-scripts/issues/323.

**Comment by shyammarjit** [comment]
> Hey @shyammarjit,
> 
> Can you add a code snippet that shows how the code is currently failing?

From https://github.com/huggingface/diffusers/issues/5016
```
export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
export OUTPUT_DIR= <put output dir>
export INSTANCE_DIR= <put dataset path>

accelerate launch train_dreambooth_lora_sdxl.py
--pretrained_model_name_or_path=$MODEL_NAME \
--instance_data_dir=$INSTANCE_DIR \
--output_dir=$OUTPUT_DIR \
--mixed_precision="fp16" \
--instance_prompt="a photo of sks {subject}" \
--resolution=1024 \
--train_batch_size=1 \
--gradient_accumulation_steps=4 \
--learning_rate=1e-4 \
--lr_scheduler="constant" \
--lr_warmup_steps=0 \
--max_train_steps=1000 \
--seed="0" \
--use_8bit_adam \
--gradient_checkpointing \
--enable_xformers_memory_efficient_attention \
--train_text_encoder \

```

**Comment by patrickvonplaten** [comment]
> > Hey @shyammarjit,
> > Can you add a code snippet that shows how the code is currently failing?
> 
> From #5016
> 
> ```
> export MODEL_NAME="stabilityai/stable-diffusion-xl-base-1.0"
> export OUTPUT_DIR= <put output dir>
> export INSTANCE_DIR= <put dataset path>
> 
> accelerate launch train_dreambooth_lora_sdxl.py
> --pretrained_model_name_or_path=$MODEL_NAME \
> --instance_data_dir=$INSTANCE_DIR \
> --output_dir=$OUTPUT_DIR \
> --mixed_precision="fp16" \
> --instance_prompt="a photo of sks {subject}" \
> --resolution=1024 \
> --train_batch_size=1 \
> --gradient_accumulation_steps=4 \
> --learning_rate=1e-4 \
> --lr_scheduler="constant" \
> --lr_warmup_steps=0 \
> --max_train_steps=1000 \
> --seed="0" \
> --use_8bit_adam \
> --gradient_checkpointing \
> --enable_xformers_memory_efficient_attention \
> --train_text_encoder \
> ```

@sayakpaul do you have time to check this in more detail?

**Comment by sayakpaul** [comment]
@patrickvonplaten see https://github.com/kohya-ss/sd-scripts/issues/323#issuecomment-1485073421. It's apparently very specific to gradient checkpointing. Off top of my head, I think it's because of the way gradient recompute chain formulated during the backward pass. 

I think we could just condition the proposed changes on the enablement of gradient checkpointing here.

**Comment by shyammarjit** [comment]
After adding these two lines still there is one bug 
@sayakpaul  and @patrickvonplaten please have a look into https://github.com/huggingface/diffusers/issues/5016#issuecomment-1722636558 this issue.

**Review by sayakpaul — APPROVED** [review]
